File: 백준/Gold/6987. 월드컵/월드컵.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static final int COUNTRY = 6;
    static int[][] score = new int[COUNTRY][3];
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();
        for(int f = 0; f < 4; f++) {
            st = new StringTokenizer(br.readLine());

            int totalWin = 0, totalDraw = 0, totalLoss = 0;
            boolean flag = true;
            for(int c = 0; c < COUNTRY; c++) {
                int win = Integer.parseInt(st.nextToken());
                int draw = Integer.parseInt(st.nextToken());
                int loss = Integer.parseInt(st.nextToken());

                score[c][0] = win;
                score[c][1] = draw;
                score[c][2] = loss;

                totalWin += win;
                totalDraw += draw;
                totalLoss += loss;

                if(win + draw + loss != 5) flag = false;
            }

            if(totalWin != totalLoss || totalDraw % 2 != 0) flag = false;
            if(flag) {
                flag = dfs(0, 1);
            }

            sb.append((flag ? "1 " : "0 "));
        }

        System.out.println(sb);
    }

    static boolean dfs(int country, int target) {
        if(country >= COUNTRY) return true;
        if(target >= COUNTRY) return dfs(country + 1, country + 2);

        for(int i = 0; i < 3; i++) {
            if(score[country][0 + i] > 0 && score[target][2 - i] > 0) {
                score[country][0 + i]--;
                score[target][2 - i]--;
                if(dfs(country, target + 1)) return true;
                score[country][0 + i]++;
                score[target][2 - i]++;
            }
        }

        return false;
    }
}

File: 백준/Gold/6987. 월드컵/월드컵.cc
#include <iostream>
#define COUNTRY 6
using namespace std;

int board[COUNTRY][3] = { 0 };
int tmp[COUNTRY][3] = { 0 };
int result[4] = { 0 };

bool is_possible(int country, int target);

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);

	for (int f = 0; f < 4; f++) {
		bool flag = true;
		int total_win = 0, total_loss = 0, total_draw = 0;
		for (int country = 0; country < COUNTRY; country++) {
			int win, draw, loss;
			cin >> win >> draw >> loss;

			board[country][0] = win;
			board[country][1] = draw;
			board[country][2] = loss;

			if (win + draw + loss != 5) flag = false;
			total_win += win;
			total_loss += loss;
			total_draw += draw;
		}

		if (total_win != total_loss || total_draw % 2 != 0) flag = false;

		if (flag) {
			flag = is_possible(0, 1);
		}

		result[f] = flag ? 1 : 0;
	}

	for (int i = 0; i < 4; i++) {
		cout << result[i] << " ";
	}
}

bool is_possible(int country, int target) {
	if (country >= COUNTRY) {
		return true;
	}
	if (target >= COUNTRY) {
		return is_possible(country + 1, country + 2);
	}

	for (int i = 0; i < 3; i++) {
		if (board[country][0 + i] > 0 && board[target][2 - i] > 0) {
			board[country][0 + i]--;
			board[target][2 - i]--;
			if(is_possible(country, target + 1)) return true;
			board[country][0 + i]++;
			board[target][2 - i]++;
		}
	}
	return false;
}